File: Memoization/509_FibonacciNumber.cs
// 509. Fibonacci Number
// Tags: memoization, dynamic programming
//
// Note: This has 2 solutions.

// Regular solution
public class Solution 
{   
    public int Fib(int n) 
    {
        if(n == 0)
            return 0;
        if(n == 1)
            return 1;
        
        return Fib(n - 1) + Fib(n - 2);
    }
}

// Memoized solution
public class Solution 
{
    private int[] _cache;
    
    public int Fib(int n) 
    {
        if(n == 0)
            return 0;
        if(n == 1)
            return 1;
        if(_cache == null)
            _cache = new int[n];
        
        // Since C# arrays initialize to 0, it is safe
        // to assume that an array value of 0 is
        // uninitialized.
        if(_cache[n - 1] == 0)
            _cache[n - 1] = Fib(n - 1) + Fib(n - 2);
            
        return _cache[n - 1];
    }
}

File: Memoization/509_FibonacciNumber.rs
// 509. Fibonacci Number
// Tags: memoization, dynamic programming
//

impl Solution {
    pub fn fib(n: i32) -> i32 {

        let mut cache:[i32; 30] = [0; 30];

        fn fib_seq(m: usize, cache: &mut [i32; 30]) -> i32 {
            if m == 0 {
                return 0;
            }
            if m == 1 {
                return 1;
            }
            
            if cache[m - 1] == 0 {
                cache[m - 1] = fib_seq((m - 1), cache) + fib_seq((m - 2), cache);
            }

            cache[m - 1]
        }

        fib_seq((n as usize), &mut cache)
    }
}